solvers/mosek/mosekbackend.cc

```diff
@@ -378,135 +378,174 @@ double MosekBackend::MIPGapAbs() {
     }
-  }*/
+  }
   return cons;
 }
 
 void MosekBackend::VarStatii(ArrayRef<int> vst) {
-  // TODO
-  int index[1];
+  // BasicStatus enum values: from 0 to 6: none, bas, sup, low, upp, equ, btw
   std::vector<int> stt(vst.data(), vst.data() + vst.size());
-  /*
-  for (auto j = stt.size(); j--; ) {
-    auto& s = stt[j];
-    switch ((BasicStatus)s) {
-    case BasicStatus::bas:
-      s = MOSEK_BASIS_BASIC;
-      break;
-    case BasicStatus::low:
-      s = MOSEK_BASIS_LOWER;
-      break;
-    case BasicStatus::equ:
-      s = MOSEK_BASIS_FIXED;
-      break;
-    case BasicStatus::upp:
-      s = MOSEK_BASIS_UPPER;
-      break;
-    case BasicStatus::sup:
-    case BasicStatus::btw:
-      s = MOSEK_BASIS_SUPERBASIC;
-      break;
-    case BasicStatus::none:
-      /// 'none' is assigned to new variables. Compute low/upp/sup:
-      /// Depending on where 0.0 is between bounds
-      double lb, ub;
-      index[0] = (int)j;
-      if(!MOSEK_GetColInfo(lp(), MOSEK_DBLINFO_LB, 1, index, &lb) && 
-        !MOSEK_GetColInfo(lp(), MOSEK_DBLINFO_UB, 1, index, &ub))
-      { 
-        if (lb >= -1e-6)
-          s = -1;
-        else if (ub <= 1e-6)
-          s = -2;
-        else
-          s = -3;  // or, leave at 0?
-      }
-      break;
-    default:
-      MP_RAISE(fmt::format("Unknown AMPL var status value: {}", s));
+  int *skx = (int *) MSK_calloctask(lp(), stt.size(), sizeof(int));
+
+  // TODO: remove this if we are sure that it is the same.
+  MSKint32t numvar;
+  MOSEK_CCALL(MSK_getnumvar(lp(), &numvar));
+  assert(numvar == vst.size());
+
+  for (auto j = 0; j < stt.size(); j++)
+  {
+    skx[j] = MSK_SK_UNK;
+    switch ((BasicStatus)stt[j])
+    {
+      case BasicStatus::bas:
+        skx[j] = MSK_SK_BAS;
+        break;
+      case BasicStatus::low:
+        skx[j] = MSK_SK_LOW;
+        break;
+      case BasicStatus::equ:
+        skx[j] = MSK_SK_FIX;
+        break;
+      case BasicStatus::upp:
+        skx[j] = MSK_SK_UPR;
+        break;
+      case BasicStatus::sup:
+      case BasicStatus::btw:
+        skx[j] = MSK_SK_SUPBAS;
+        break;
+      case BasicStatus::none:
+        /// 'none' is assigned to new variables (added after solving).
+        /// Compute low/upp/fix/sup, depending on where initial value 0.0 is between the bounds
+        MSKboundkeye bk;
+        MSKrealt bl, bu, tolx;
+        MOSEK_CCALL(MSK_getvarbound(lp(), (MSKint32t)j, &bk, &bl, &bu));
+        MOSEK_CCALL(MSK_getdouparam(lp(), MSK_DPAR_DATA_TOL_X, &tolx));
+
+        if (0.0 < bl + tolx) skx[j] = MSK_SK_LOW;
+        else if (bu - tolx < 0.0) skx[j] = MSK_SK_UPR;
+        else if (bu - bl < tolx) skx[j] = MSK_SK_FIX;
+        else skx[j] = MSK_SK_SUPBAS;
+
+        break;
+      default:
+        MP_RAISE(fmt::format("Unknown AMPL var status value: {}", stt[j]));
     }
   }
-  MOSEK_SetBasis(lp(), stt.data(), NULL);
-  */
+
+  // Input the variable status keys
+  MOSEK_CCALL(MSK_putskx(lp(), MSK_SOL_BAS, (MSKstakeye *)skx));
```

Comment:
Is the memory of *skx lost? I understand it should be deallocated.

Reply:
You are right, it uses vector now.